subprojects/core/src/main/java/org/gradle/invocation/EagerLifecycleExecutor.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.invocation;
+
+import org.gradle.api.IsolatedAction;
+import org.gradle.api.Project;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
+
+@ServiceScope(Scope.Build.class)
+public interface EagerLifecycleExecutor {
```

Comment:
💭  I wonder if the "eager" naming makes sense here. The previous `gradle.allproject` callbacks were eager, because they executed unconditionally and eagerly *before* the actual project evaluation.

The new behavior we introduce in this PR is more "reactive". It is also not eager at all, if no mutable access occurs.

I don't have good suggestions for the name, and the potential rename is not a blocker for merging, but it's something to think about

Reply:
I think this particular interface does not actually care when the callbacks are executed. It is an external concern.

So maybe it is easier to call based on what it does: `BeforeProjectCallbackExecuter`.

Reply:
Now it's `LifecycleActionExecutor`

Reply:
That's much nicer IMO, thanks!

As a further naming improvement, consider calling it either `GradleLifecycleActionExecuter` (that's where users register those currently) or `ProjectLifecycleActionExecutor` (that's whose lifecycle these callbacks actually affect).